pkg/util/ip_reservation.go

```diff
@@ -0,0 +1,148 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"fmt"
+	"net"
+)
+
+const (
+	incrementStep29IPBlock = byte(8)
+	reservationBlockSize   = "/29"
+)
+
+// IPAllocator struct maintains the current list of reserved /29 IPs pending a write operation in the cloud service. An IP Range is freed up from this list when it is available at the cloud provider end
+// As the create instance operation is in parallel, it is possible that an IP range was reserved by the service instance but is not updated in the cloud instance. The reserved list stores such IP ranges
+type IPAllocator struct {
+	reserved map[string]bool
+}
+
+// NewIPAllocator is the constructor to initialize the IPAllocator object
+// Argument reserved map[string]bool is a set of currently reserved IPs for which the CreateInstance operation is not completed
+func NewIPAllocator(reserved map[string]bool) *IPAllocator {
+	return &IPAllocator{
+		reserved: reserved,
+	}
+}
+
+// Reserve function adds a particular IP block in the reserve pool
+// Argument ipBlock string is an IPV4 range which needs to be reserved
+func (ipAllocator *IPAllocator) Reserve(ipV4Range string) {
+	ipAllocator.reserved[ipV4Range] = true
+}
+
+// Unreserve function frees up a reserved IP
+// Argument ipBlock string is an IPV4 range which needs to be unreserved
+func (ipAllocator *IPAllocator) Unreserve(ipV4Range string) {
+	delete(ipAllocator.reserved, ipV4Range)
+}
+
+// ValidateCIDR function validates whether a particular cidr is a valid IP range
+// Argument cidr string is is a CIDR range that needs to be validated
+func (ipAllocator *IPAllocator) ValidateCIDR(cidr string) bool {
+	_, _, err := net.ParseCIDR(cidr)
+	return err == nil
+}
+
+// GetUnreservedIPBlock returns an unreserved /29 IP block.
+// cidr: Provided cidr address in which we need to look for an unreserved /29 IP Block
+// reservedIPRanges: All the Reserved IP ranges present in the cloud provider
+// All the Reserved IP ranges not present in the cloud provider are extracted from the reserved list in the IPAllocator
+// Finally a final reserved list is created by merging these two lists
+// Potential error cases: 1) No /29 Block in the CIDR is unreserved
+// 2) Parsing the CIDR resulted in an error
+func (ipAllocator *IPAllocator) GetUnreservedIPBlock(cidr string, reservedIPRanges map[string]bool) (string, error) {
+	// Merging the reserved lists at the cloud provider end and the service end to form a final reserved list
+	for reservedIPRange := range ipAllocator.reserved {
+		reservedIPRanges[reservedIPRange] = true
```

Comment:
This should create and modify a new map instead of modifying the user's map.  Also does this need to be a map? We seem to just be iterating over it, so it could be a slice.

Reply:
Its implementation of a Set using a Map